urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/framework/ExecFromData.java

```diff
@@ -0,0 +1,393 @@
+
+            //Exceeded time
+            long now = System.currentTimeMillis();
+            long spent = now - then;
+
+            if(spent > PTSPConstants.TIME_ACTION_DISQ)
+            {
+                actionToExecute = 0;
+                System.out.println("Controller disqualified. Time exceeded: " + (spent - PTSPConstants.TIME_ACTION_DISQ));
+                m_game.abort();
+
+            }else{
+
+                if(spent > PTSPConstants.ACTION_TIME_MS)
+                    actionToExecute = 0;
+                m_game.tick(actionToExecute);
+            }
+
+            //And paint everything.
+            if(visual)
+            {
+                m_view.repaint();
+                if(m_game.getTotalTime() == 1)
+                    waitStep(m_warmUpTime);
+            }
+        }
+
+        //Print results.
+        if(m_verbose)
+            m_game.printResults();
+
+        //And save the route, if requested:
+        if(m_writeOutput)
+            m_game.saveRoute();
+    }
+
+
+    /**
+     * Runs a game from data, receiving the controller by parameter. Returns the game when finished.
+     * @param controller  Controller to run. <b> Important: </b> It must be created from outside: it does not
+     *                    check that the initialization time is respected!
+     * @param visual      Whether to show graphics or not.
+     * @param accelerated Whether the game must be accelerated or not.
+     * @param map         Data grid with the map information (free and obstacle spaces).
+     * @param startingPoint Starting point of the ship.
+     * @param wayPoints   List with the positions of the waypoints.
+     * @return the instance of the game played.
+     */
+    public static Game runGame(Controller controller, boolean visual, boolean accelerated,
+                               char[][] map, Vector2d startingPoint, LinkedList<Vector2d> wayPoints, LinkedList<Vector2d> fuelTanks)
+    {
+        m_controller = controller;
+        controller.reset();
+
+        //Create the game instance.
+        m_game = new Game(map, startingPoint, wayPoints,fuelTanks);
+        m_game.go();
+        m_game.getShip().setStarted(true);
+
+        JEasyFrame frame = null;
+        if(visual)
+        {
+            //View of the game, if applicable.
+            m_view = new PTSPView(m_game, m_game.getMapSize(), m_game.getMap(), m_game.getShip(), m_controller);
+            frame = new JEasyFrame(m_view, "PTSP-Game: " + m_controller.toString());
+        }
+
+        while(!m_game.isEnded())
+        {
+            //When the result is expected:
+            long due = System.currentTimeMillis()+PTSPConstants.ACTION_TIME_MS;
+
+            //Advance the game.
+            m_game.tick(m_controller.getAction(m_game.getCopy(), due));
+
+            if(!accelerated)
+                waitStep(10);
+
+            //And paint everything.
+            if(visual)
+            {
+                m_view.repaint();
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 127:143, 184:200, 243:259.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.